### ISSUE TITLE
spatial clustering run time

### ISSUE THREAD

**Issue by jeffreyhanson** [NCC-CNC/wheretowork#139]
Joe had this comment about the app:

> Spatial clustering much slower (not surprising)? It took about 15 min to run with 50% clustering, where other runs were very quick (> 1 min).

Yeah, it would be great to reduce run time for spatial clustering. The prioritizations are generated using a multi-objective approach that comprises two sequential optimization runs (one to find a solution based on goals + weights, and a second one that adds spatial clustering). One approach to reduce run time for spatial clustering would involve (1) using a starting solutions to speed up the second optimization run, and (2) using a time limit for the second optimization run. This is trivial for Gurobi - but I haven't worked out how to do this for CBC yet.

**Comment by jeffreyhanson** [comment]
This issue is also simialr to run time stuff talked about in https://github.com/NCC-CNC/wheretowork/issues/94

**Comment by ricschuster** [comment]
I think this is not critical for now, but we could try to figure this out in the next iteration.

We (NCC) could eventually also approach Gurobi and ask them if we could get a free license from them, given our line of work, but I'm not sure how far that is off and how realistic it actually is.

**Comment by jeffreyhanson** [comment]
I've been playing around with various methods to speed up the optimization process. It seems that these methods do have some success - but solving the case-study problem with no selected Includes will still take ages. These methods involve modifying the secondary optimization run used to promote spatial clustering to:

1. Use the initial solution based on just cost + features + weights as a starting solution for the run. The idea here is to help speed up the solver by giving it a decent starting solution.
2. Calculating planning unit importance scores (based on range-weighted rarity) and locking in planning units selected in the initial solution that have high importance scores (i.e. above median). The idea here is to help speed up the solver by fixing variables that are likely to be selected in the second solution that promotes spatial clustering.
3. Instead of using boundary penalties, use connectivity penalties that parametrize connectivity between pairs of planning units based on their range-weighted rarity scores (note that this is not just limited to planning units selected in initial run). The idea here is to help speed up the solver by making the objective coefficients more variable (but also somewhat sensible).
4. Setting pair-wise connectivity values to zero if they are below the median. The idea here is to help speed up the solver by reducing the problem size -- and this is acheived by "removing" connections between pairs of planning units that are unlikely to be selected in the solution anyway (because neither are important).

In addition to thse four methods, I also played around with setting connectivity penalties based on the weights. This worked fairly well when using the "Protected Area distance" weight -- but this method won't work for problems where there are no weights selected (and is less likely to to work well for binary weights). So, I haven't really looked into it further.

**Comment by jeffreyhanson** [comment]
@josephrbennett, I've just pushed a new version of the app to https://ncc.carleton.ca/ with the speed improvements mentioned above. When you get a chance, could you please take a look and see what you think?

**Comment by ricschuster** [comment]
Thanks Jeff!
Which points did you incorporate? Looks like not 4., but how about the other ones?

**Comment by jeffreyhanson** [comment]
Yeah, that's right, I've incorperated all 6 except for 4 into 53f72dd.
